### ISSUE TITLE
More flexibility in modifiers for key bindings

### ISSUE THREAD

**Issue by mafredri** [kasper/phoenix#143]
Before macOS Sierra I used Seil to bind Caps Lock to F19 and Karabiner to bind F19 to Ctrl+Opt+Cmd. This allowed me to use Caps Lock as a modifier key. 

Currently only Karabiner-Essentials works on Sierra which provides a lightweight key-to-key 1-to-1 mapping (for now). 

This lead me to wonder if there's a reason for the limitation that only Cmd, Opt, Ctrl, Shift can work as modifier keys?

Right now I'm using a keybinding on f19 that enables all my key bindings without modifiers and keeps them active while the key is pressed down (using key repeat). FYI this works OK for my purposes.

My thought is that couldn't any key work as a modifier? 

Sorry for the long intro, just wanted to provide proper context.

**Comment by kasper** [comment]
Well, at least the Carbon hot key API has strict set of modifiers and Phoenix supports all of them. As far as I know, this is the only way to exclusively bind a key combination to an app (i.e. the event will not reach any other apps).

You can also listen to all key events, but those will go through all apps. For example, Phoenix could do something with it, but if you had your cursor on a text field, the event would cause the text field to do an action. (This would pretty much be a key logger.)

Finally, there’s a concept of tapping to events, which as far as I know, can be used to intercept events and prevent them from reaching other apps. I’m not really familiar with it. Frankly, I don’t know how Seil or Karabiner achieve what you described.

TLDR: With the current implementation, no. You can bind to a strict set of modifiers (or no modifiers all, which gives you access to a single key character).

**Comment by mafredri** [comment]
Ok, thanks for clarifying the reason for the current implementation, makes perfect sense and considering the other options, I don't think this is worth a hacky solution.

I believe Seil/Karabiner uses kernel extensions for at least some of this functionality.

With my current binding of F19 without modifiers I have 99% of what I want anyway. The only thing I can't achieve is immediate unbinding of keys when F19 is released (500ms delay on initial keyrepeat and 100ms if it's "repeating"). Haven't used this long enough to know if that even is a problem 😁.

**Comment by kasper** [comment]
Ok! :smiley: I guess you could achieve any set of modifiers by implementing a hot binding shim using a single key as an entry point and then binding to a set of keys. Not worth adding that to the core though.
